src/org/javarosa/model/xform/CompactSerializingVisitor.java

```diff
@@ -0,0 +1,210 @@
+     * .DataModelTree)
+     */
+    public void visit(FormInstance tree) {
+
+        TreeElement root = tree.getRoot();
+
+        delimiter = root.getAttributeValue(NAMESPACE_ODK, "delimiter");
+        if (delimiter == null) {
+            // for the spelling-impaired...
+            delimiter = root.getAttributeValue(NAMESPACE_ODK, "delimeter");
+        }
+        String prefix = root.getAttributeValue(NAMESPACE_ODK, "prefix");
+
+        delimiter = (delimiter != null) ? delimiter : " ";
+        prefix = (prefix != null) ? prefix : " ";
+
+        resultText = prefix.concat(delimiter);
+
+        // serialize each node (and its children) to get its answers
+        resultText = serializeTreeToString(root);
+    }
+
+    private String serializeTreeToString(TreeElement root) {
+        StringBuilder sb = new StringBuilder();
+        serializeTree(root, sb);
+        return sb.toString().trim();
+    }
+
+    private void serializeTree(TreeElement root, StringBuilder sb) {
+        for (int j = 0; j < root.getNumChildren(); j++) {
+            TreeElement treeElement = root.getChildAt(j);
+            if (treeElement.isLeaf() && treeElement.getAttribute("", "tag") != null) {
+                String result = serializeNode(treeElement);
+                if (result != null) {
+                    sb.append(result);
+                }
+            } else {
+                serializeTree(treeElement, sb);
+            }
+        }
+    }
+
+    public String serializeNode(TreeElement instanceNode) {
+        StringBuilder stringBuilder = new StringBuilder();
+        // don't serialize template nodes or non-relevant nodes
+        if (!instanceNode.isRelevant()
+            || instanceNode.getMult() == TreeReference.INDEX_TEMPLATE) {
+            return null;
+        }
+
+        if (instanceNode.getValue() != null) {
+            Object serializedAnswer = serializer.serializeAnswerData(
+                instanceNode.getValue(), instanceNode.getDataType());
+
+            if (serializedAnswer instanceof Element) {
+                // DON"T handle this.
+                throw new RuntimeException("Can't handle serialized output for "
+                    + instanceNode.getValue().toString() + ", "
+                    + serializedAnswer);
+            } else if (serializedAnswer instanceof String) {
+                Element element = new Element();
+                element.addChild(Node.TEXT, serializedAnswer);
+
+                String tag = instanceNode.getAttributeValue("", "tag");
+
+                stringBuilder.append(tag);
+
+                stringBuilder.append(delimiter);
+
+                for (int k = 0; k < element.getChildCount(); k++) {
+                    stringBuilder.append(element.getChild(k).toString().
+                        replace("\\", "\\\\")
+                        .replace(delimiter, "\\" + delimiter)
+                        .replace(tag, "\\" + tag));
```

Comment:
I'm not sure about escaping tags. I expect most recipients will tokenize incoming messages using the delimiter and consider values in even positions to be tags and values in odd positions to be corresponding answer values. This kind of escaping makes processing much more complex.

Take this simple example where a form designer is using alphabetic designations as tags (a, b, c):
```
<fname odk:tag="a">Alabama</fname>
```
The serialized version given a delimiter of " " would be `a Al\ab\am\a`. 

I believe this is something @ekigamba asked for. Could you please explain in more depth why you would like it? Can we launch an initial version without it and reconsider if needed?

Reply:
Oh yes, that's a really good point and example. I don't really have any feelings towards it. So I think we can remove it for now.

Reply:
Yes, let's. Once it's in it's really hard to take away. If there's user demand for it we can draft up a rationale and consider implications.